src/kernl/optimizer/cuda_graph.py

```diff
@@ -12,48 +12,74 @@
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
 #
-from collections import defaultdict, deque
+import math
 from typing import Callable, Union
 
 import torch
-import triton
 from torch._dynamo import utils as dynamo_utils
 from torch._inductor.compile_fx import cudagraphify_impl
 from torch._subclasses import FakeTensor
 
+from kernl.optimizer.pool_cuda_graphs import CudaGraphPool, get_aligned_size
 
-static_inputs_pool: dict[(int, torch.dtype), deque[torch.Tensor]] = defaultdict(deque)
 
+static_inputs_pool = []
 
-def get_static_inputs(model_inputs: list[torch.Tensor]) -> list[torch.Tensor]:
-    """
-    Copy input tensors to a managed pool of tensors (to limit memory footprint).
-    All tensors need to be declared as static inputs to CUDA graph.
 
-    @param model_inputs: list of inputs
-    @return: list of inputs to be used in CUDA graphs
+def prepare_inputs(inputs: list[torch.Tensor], pools: list[CudaGraphPool]) -> list[torch.Tensor]:
+    """
+    Copy the inputs in the CUDA graphs memory pool and return tensor copies.
+    Follows a greedy bin packing algorithm (first-fit decreasing) to minimize the number of pools:
+    - sort the items in decreasing order of size ;
+    - insert them one by one into the first bin that has room for it.
+
+    :param inputs: list of tensors to copy in the pool
+    :param pools: list of available pools
+    :return: list of tensors that are copies of the inputs in the pool
     """
-    static_inputs_clone: dict[(int, torch.dtype), deque[torch.Tensor]] = defaultdict(deque)
-    for k, v in static_inputs_pool.items():
-        static_inputs_clone[k] = v.copy()
-    cuda_graph_input: list[torch.Tensor] = list()
-    for index, original_tensor in enumerate(model_inputs):
-        storage_size = triton.next_power_of_2(len(original_tensor.untyped_storage()))
-        tensor_pool_key = (storage_size, original_tensor.dtype)
-        if len(static_inputs_clone[tensor_pool_key]) > 0:
-            static_tensor = static_inputs_clone[tensor_pool_key].popleft()
-        else:
-            static_tensor = torch.empty((storage_size,), dtype=original_tensor.dtype, device=original_tensor.device)
-            static_inputs_pool[tensor_pool_key].append(static_tensor)
-
-        # storage offset should not be used below... otherwise it changes cuda address
-        before_ptr = static_tensor.data_ptr()
-        static_tensor = torch.as_strided(static_tensor, original_tensor.size(), original_tensor.stride())
-        static_tensor.copy_(original_tensor)
-        cuda_graph_input.append(static_tensor)
-        assert before_ptr == static_tensor.data_ptr(), f"should be equal: {before_ptr} != {static_tensor.data_ptr()}"
-
-    return cuda_graph_input
+    inputs_copy = list(inputs).copy()
+    # add position meta
+    for index, t in enumerate(inputs_copy):
+        t.position = index
```

Comment:
I don't think so, later the (new) order is lost because inputs are split between those that have been copied in an existing pool and the other to be associated to a pool to create. Do you see a way to avoid that?